src/material-experimental/mdc-slider/slider.ts

```diff
@@ -245,3 +373,183 @@ export class MatSlider implements AfterViewInit, OnDestroy {
+  // are already exposed so users can just listen for those events directly themselves.
+  emitChangeEvent = (value: number, thumb: Thumb): void => {};
+  emitInputEvent = (value: number, thumb: Thumb): void => {};
+  emitDragStartEvent = (value: number, thumb: Thumb): void => {
+    const input = this._delegate._getInput(thumb);
+    input.dragStart.emit({ source: input, parent: this._delegate, value });
+  }
+  emitDragEndEvent = (value: number, thumb: Thumb): void => {
+    const input = this._delegate._getInput(thumb);
+    input.dragEnd.emit({ source: input, parent: this._delegate, value });
+  }
+  registerEventHandler =
+    <K extends EventType>(evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._elementRef.nativeElement.addEventListener(evtType, handler);
+  }
+  deregisterEventHandler =
+    <K extends EventType>(evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._elementRef.nativeElement.removeEventListener(evtType, handler);
+  }
+  registerThumbEventHandler =
+    <K extends EventType>(thumb: Thumb, evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._getThumbElement(thumb).addEventListener(evtType, handler);
+  }
+  deregisterThumbEventHandler =
+    <K extends EventType>(thumb: Thumb, evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._getThumbElement(thumb).removeEventListener(evtType, handler);
+  }
+  registerInputEventHandler =
+    <K extends EventType>(thumb: Thumb, evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._getInputElement(thumb).addEventListener(evtType, handler);
+  }
+  deregisterInputEventHandler =
+    <K extends EventType>(thumb: Thumb, evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._getInputElement(thumb).removeEventListener(evtType, handler);
+  }
+  registerBodyEventHandler =
+    <K extends EventType>(evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._document.body.addEventListener(evtType, handler);
+  }
+  deregisterBodyEventHandler =
+    <K extends EventType>(evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._document.body.removeEventListener(evtType, handler);
+  }
+  registerWindowEventHandler =
+    <K extends EventType>(evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._window.addEventListener(evtType, handler);
+  }
+  deregisterWindowEventHandler =
+    <K extends EventType>(evtType: K, handler: SpecificEventListener<K>): void => {
+      this._delegate._window.removeEventListener(evtType, handler);
+  }
+}
+
+/**
+ * Ensures that there is not an invalid configuration for the slider thumb inputs.
+ */
+function _validateInputs(
+  isRange: boolean,
+  startInputElement: HTMLInputElement,
+  endInputElement: HTMLInputElement): void {
+  if (isRange) {
+    if (!startInputElement!.hasAttribute('matSliderStartThumb')) {
```

Comment:
Why is the `!` necessary on `startInputElement!`? The function signature indicates that it can't be null/undefined